routers/web/web.go

```diff
@@ -1446,17 +1446,17 @@ func registerRoutes(m *web.Router) {
 		m.Group("/contributors", func() {
 			m.Get("", repo.Contributors)
 			m.Get("/data", repo.ContributorsData)
-		})
+		}, reqRepoCodeReader)
```

Comment:
That's just questionable, why it needs to check this permission? Does the handler expose any "real code"?

Reply:
The original logic in the template needs code permission or it will display nothing.
So, I think it should have same permission checks in router, or you removed them in the menu in UI, but you can still access these pages by just editing the URL.

And for the original logic, maybe it depends on the definition of `Code`: does author info include in it?

Reply:
Actually I have no interest to touch the legacy logic at the moment because it is quite a mess (I did do some quick fixes just because these "bugs" affected end users and are potentially related to my PRs).

---

The real problem is that there is NO clear definition for the "permission" system, people just patched the code again and again, and there are more serious abuses like "issue reader / pull reader permissions are mixed (`/{type:issues|pulls}`)".

Maybe you could define these permissions clearly first before code work. The definition should be reasonable to most end users and daily usage, but not by what it literally looks.

Reply:
IMO, anything contain the information from commits (no matter from original repo or forks) or themselves should be controlled by code unit permission, as codes are managed by git, and git uses commits to record the status/changes of these codes.
I'm not sure whether it is acceptable to others. What's your opinion?

Reply:
Why they need such strict rules?

Permissions should be as loose as possible to satisfy every users for various use cases, as long as the permission design doesn't cause harm or leak sensitive data.

Reply:
In my mind:

The purpose of a permission system is to "prevent from users doing harm to a repo".

* `code` permission means the user could access the full code. Disabling it means that prevent the user from accessing the full code. The purpose should be "protect the full code from leaking"
* `pr read` means that the user could discuss and review the code change (not the full code) , approve or reject.
* `activity` means that the user could see the statistics of the repo (commits number and authors do not really leak code)


So, that's why I would like to keep the commit/files tabs on the "pr" page, and would like to make users could read the statistics of the repo as much as possible (useful and no harm)

Reply:
It is removed.